_posts/2018-02-15-continuous-deployment-concourse.md

```diff
@@ -0,0 +1,286 @@
+
+## Build images
+
+### Docker image resource definition:
+```
+- name: foobar-app-image
+  type: docker-image
+  source:
+    repository: gcr.io/foobar/foobar_app
+    username: _json_key
+    password: {{gcloud_service_key}}
+```
+
+### Build and push the image:
+```
+- put: foobar-app-image
+  params:
+    build: artifact-app/
+    cache: true
+    cache_tag: latest
+    tag: metadata/short-commit-hash
+    tag_as_latest: true
+```
+Note: 
+- the artifact-app/ volume contains a Dockerfile and dependencies to build the image 
+- the `latest` tag will be cached on disk, if we have a lot of builds this can save time!
+- The metadata is a simple task that uses a `git` Docker image to retrieve the git short commit hash:
+    ```
+    ---
+    platform: linux
+
+    image_resource:
+      type: docker-image
+      source: {repository: greenwall/git}
+
+    params:
+      TERM: xterm-256color
+
+    inputs:
+      - name: code-base
+
+    outputs:
+      - name: metadata
+
+    run:
+      path: /bin/sh
+      args:
+      - -exc
+      - |
+        GIT_SHORT_HASH=`cd code-base && git log --pretty=format:'%h' -n 1`
+        echo $GIT_SHORT_HASH > metadata/short-commit-hash
+    ```
+
+## Run unit tests on pre-built container
+We could easily reuse that image for unit testing, tests that should not depend on anything else than the code itself. Here's an example on how to do it with php-unit:
+
+```
+- task: php-unit-tests
+  config:
+    platform: linux
+    image_resource:
+      type: docker-image
+      source:
+        repository: gcr.io/foobar/foobar_app
+        username: _json_key
+        password: {{gcloud_service_key}}
+    run:
+      path: sh
+      args:
+      - -exc
+      - |
+        /var/www/vendor/phpunit/phpunit/phpunit --testsuite=unit --bootstrap /var/www/includes/tests/bootstrap.php --configuration /var/www/includes/tests/phpunit-conf.xml
+  on_failure:
+    put: git-pr
+    params:
+      context: php-unit-tests
+      path: git-pr
+      status: failure
+- put: git-pr
+  params:
+    context: php-unit-tests
+    path: git-pr
+    status: success
+```
+
+
+## Continous deployment on staging/production
+### Tooling
+To manage releases on Kubernetes we use [Helm](http://helm.sh). This tool also require kubectl to be properly configured and since we are using GKE we need the [gcloud](https://cloud.google.com/sdk/downloads) cli. These tools do not exist on Concourse, they are too specific, this is why I built an [alpine-based image containing all these tools](https://github.com/baptiste-bonnaudet/gcloud-kubectl-helm).
+
+### Deploy and rollback on failure
+
+The `deploy-staging` job in the full pipeline below takes a helm charts repository as input, use the gcloud-kubectl-helm Docker image as a task to deploy our application to staging. We use the helm test feature to test that our release was successful before failing. And in case of failure it will attempt a rollback. 
```

Comment:
- `The deploy-staging job in the full pipeline below takes a helm charts` is clunky and should be reworded to something like `In the pipeline below, the deploy staging job [...]`
- `gcloud-kubectl-helm` is a programmatic resource and should be wrapped in backticks.
- `We use the helm test feature to test that our`, avoid using the same word twice. Replace `to test` with `to ensure`.